Add Status orbs plugin

TheStonedTurtle opened the pull request:
Combines the runenergy & regenmeter into a single plugin and adds dynamic heart functionality to the  plugin.

Adds `ImageUtil.getImageSpritePixels()` and modifies existing plugins to use it.
Adds `Client.getCurrentAffliction()` and modifies StatusBars to use it.

TheStonedTurtle commented:
Gif showing that the settings are working:
![2018-11-16_13-32-28](https://user-images.githubusercontent.com/29030969/48648456-394c2500-e9a4-11e8-8838-754c087b5bc1.gif)

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/interfacestyles/InterfaceStylesPlugin.java:
```diff
@@ -209,7 +209,7 @@ private SpritePixels getFileSpritePixels(String file, String subfolder)
 	}
 
-	private SpritePixels getImageSpritePixels(BufferedImage image)
+	public static SpritePixels getImageSpritePixels(BufferedImage image, Client client)
```
Don't reference this from another plugin

TheStonedTurtle replied:
So its better to copy and paste the exact code?

raiyni replied:
`banktags.tabs.TabSprites` has the same method. They would be moved to something like `ImageUtils`

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsOverlay.java:
```diff
@@ -50,8 +58,9 @@
 	private final Client client;
-	private RegenMeterPlugin plugin;
-	private RegenMeterConfig config;
+	private StatusOrbsPlugin plugin;
```
private final

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsOverlay.java:
```diff
@@ -50,8 +58,9 @@
-	private RegenMeterPlugin plugin;
-	private RegenMeterConfig config;
+	private StatusOrbsPlugin plugin;
+	private StatusOrbsConfig config;
```
private final

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsOverlay.java:
```diff
@@ -50,8 +58,9 @@
-	private RegenMeterConfig config;
+	private StatusOrbsPlugin plugin;
+	private StatusOrbsConfig config;
+	private TooltipManager tooltipManager;
```
private final

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsConfig.java:
```diff
@@ -22,16 +22,28 @@
+public interface StatusOrbsConfig extends Config
 {
 	@ConfigItem(
+		position = 0,
```
The position values on these configs are unnecessary

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsConfig.java:
```diff
@@ -22,16 +22,28 @@
 	@ConfigItem(
+		position = 0,
+		keyName = "dynamicHpHeart",
+		name = "Dynamic HP Heart",
```
Be consistent with the other config options

`Dynamic hitpoints heart`

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsConfig.java:
```diff
@@ -50,11 +73,13 @@ default boolean showSpecial()
-	default boolean showWhenNoChange()
+		position = 4,
+		keyName = "replaceOrbText",
+		name = "Replace Run orb text with run time left",
```
`Replace run orb text with run time left`

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -93,67 +106,224 @@
 	protected void startUp() throws Exception
 	{
-		overlayManager.add(energyOverlay);
+		overlayManager.add(overlay);
```
You should add a migrate method that sets the new keys to their old values (if it exists) and then deletes the old key.

raiyni commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -93,67 +106,224 @@
+
+		BufferedImage old = currentHeart;
+
+		currentHeart = isVenomed ? HEART_VENOM : isPoisoned ? HEART_POISON : isDiseased ? HEART_DISEASE : HEART_NORMAL;
```
Personal preference, a triple ternary shouldn't be used.

raiyni commented on runelite-client/src/main/java/net/runelite/client/util/ImageUtil.java:
```diff
@@ -407,4 +412,28 @@ private static BufferedImage offset(final BufferedImage image, final float[] sca
+	 * @param client Current client instance
+	 * @return       The buffered image as a sprite image
+	 */
+	public static SpritePixels getImageSpritePixels(BufferedImage image, Client client)
```
this is also used in bank tag tabs, so you can use this there too

TheStonedTurtle replied:
`client.getSpriteOverrides().putAll(TabSprites.toMap(client));`
I already updated `TabSprites` so this should be good :)

TheStonedTurtle commented:
Found out you can be both Poisoned/Venomed and Diseased. Added hearts to match the orb background color, some people on discord were saying to just Prioritize the Venom/Poison hearts. I'll revert if this is the preferred way to go.

Venom Disease
![https://i.imgur.com/2L2fpOJ.png](https://i.imgur.com/2L2fpOJ.png)

Poison Disease
![https://i.imgur.com/1qp7v7k.png](https://i.imgur.com/1qp7v7k.png)

Nightfirecat commented:
Made some new images for the sprite replacement; notably chose not to add a black border since that looked less natural, and adjusted hue/lightness in photoshop instead.

For reference, these are the values I used to get these images, based on the default red HP heart:

* Disease: +35 hue
* Poison: +100 hue, -5 lightness
* Venom: +125 hue, -45 lightness

![new-disease](https://user-images.githubusercontent.com/2199511/48665039-e788c500-ea9f-11e8-9b73-9874110bbc75.png)
![new-poison](https://user-images.githubusercontent.com/2199511/48665040-e788c500-ea9f-11e8-84de-c505113cf32e.png)
![new-venom](https://user-images.githubusercontent.com/2199511/48665042-e8215b80-ea9f-11e8-8b81-d34838b54db3.png)
![new-poison-disease](https://user-images.githubusercontent.com/2199511/48665041-e788c500-ea9f-11e8-8336-05e36443e43c.png)
![new-venom-disease](https://user-images.githubusercontent.com/2199511/48665043-e8215b80-ea9f-11e8-84f7-85137d4018e3.png)

TheStonedTurtle commented:
New Images: https://i.imgur.com/aCcF1T3.png

Nightfirecat commented:
I'm inclined to say that given both the rarity of disease, and the weird look of split hearts, it would probably just be better to not use them, and to prioritize poison/venom for coloring those hearts.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -243,4 +443,28 @@ int getEstimatedRecoverTimeRemaining()
+	/**
+	 * Migrates configs from runenergy and regenmeter to this plugin and deletes the old config values.
+	 */
+	private void migrateConfigs()
```
This should be annotated `@Deprecated`, and have a note regarding removing it after some time. See [`FarmingTracker.migrateConfiguration()` from the time tracking PR](https://github.com/runelite/runelite/pull/4510/files#diff-98debac5d25615bb4494938070777f37R237) or [`GroundMarkerPlugin.setColorConfiguration()` from my colored ground markers PR](https://github.com/runelite/runelite/pull/5890/files#diff-c65443644d54b889eb2a4afdcaa07550R255).

TheStonedTurtle commented:
Removed Split hearts, Fixed a config check, added deprecated on the config migration functions and changed `getCurrentAffliction()` to `getCurrentAfflictions()` which returns an EnumSet instead.

![https://i.imgur.com/a9szspr.png](https://i.imgur.com/a9szspr.png)
![https://i.imgur.com/4H4c5lp.png](https://i.imgur.com/4H4c5lp.png)

LMK if there's anything else you find that needs changing.

Nightfirecat commented:
Now that this should be feature-complete, can the history be cleaned up back into the original couple of commits it should be?

TheStonedTurtle commented:
I've cleaned up the commit history, let me know if I should squash any of the new commits together.

Nightfirecat commented on runelite-api/src/main/java/net/runelite/api/Client.java:
```diff
@@ -1575,4 +1575,10 @@
+
+	/**
+	 * Returns the Afflictions currently affecting the local player
+	 * @return
```
Better to just leave this out if you're not adding a description of it. Realistically though, the whole javadoc for this entry could be:

```
/**
 * Gets afflictions affecting the local player
 * @return A set of {@link Affliction}s affecting the local player.
 */
```

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsConfig.java:
```diff
@@ -0,0 +1,43 @@
+	{
+		return true;
+	}
+}
```
Add newline at eof I guess

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2018 TheStonedTurtle <https://github.com/TheStonedTurtle>
+ * Regen Meter Contributors:
```
These contributor entries should (a) not be separated by type, and (b) not all added in this commit; they should be added in the later commits where the other plugins are merged into this one.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -0,0 +1,184 @@
+	description = "Settings for the Minimap orbs (HP, Run, & Special Attack)",
+	tags = {"hp", "heart", "minimap", "run", "energy"}
+)
+@Slf4j
```
This annotation appears to be unused.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -0,0 +1,184 @@
+	@Subscribe
+	private void onVarbitChanged(VarbitChanged e)
+	{
+		// Dynamic HP Heart
```
This comment doesn't seem very helpful, given `config.dynamicHpHeart()` is pretty self-explanatory.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -0,0 +1,184 @@
+		EnumSet<Affliction> old = currentAfflictions;
+		currentAfflictions = client.getCurrentAfflictions();
+
+		if (old != currentAfflictions)
```
Shouldn't this be `!old.equals(currentAfflictions)`?

TheStonedTurtle commented:
Requested changes have been made to the respective commits

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -126,6 +160,61 @@ private void onVarbitChanged(VarbitChanged e)
+		{
+			ticksSinceHPRegen = 0;
+		}
+		wasRapidHeal = isRapidHeal;
```
This can move inside the if statement above.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -58,6 +70,39 @@
 )
 public class StatusOrbsPlugin extends Plugin
 {
+	private static final ImmutableSet<Integer> ALL_GRACEFUL_HOODS = ImmutableSet.of(
```
Looks like this comment is missing [from the original](https://github.com/runelite/runelite/pull/6520/commits/66e5b8aa608d0a8806fc03f66d09ae057952da73#diff-16890d9c8334c0901971acbf4f7b5de5L57)?

```
// TODO It would be nice if we have the IDs for just the equipped variants of the Graceful set items.
```

Not a big deal, but we should make sure that's not forgotten.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/statusorbs/StatusOrbsPlugin.java:
```diff
@@ -262,6 +335,102 @@ private void resetHealthIcon()
+	 * Calculates the estimated time the player can continue running for
+	 *
+	 * @param inSeconds
+	 * @return
```
Like before, this should not be left empty. Perhaps:

```
	 * @return A human-readable String of format {@code "m:ss"}.
```

TheStonedTurtle commented:
Fixed the comment and javadoc issue.
